### PR TITLE
APF concurrency isolation integration test

### PR DESCRIPTION
**Pull request by cyang49**
* Add APF concurrency utilization test

#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This is a new Kubernetes API server priority and fairness (APF) integration test. This test is designed to verify that APF regulates concurrency as expected: it should provide concurrency to priority levels (PL) as configured, and the concurrency provided to different priority levels should be isolated. In the test, we use a more demanding test client, noxu1, and a less demanding client, noxu2, to stream simple GET requests of a single Kubernetes resource object repeatedly. We expect the API server to grant all of the seats allocated for noxu1 but not exceeding the PL seat limit. For the less demanding client, we expect its requests to be served with minimal latency (i.e. little queueing overhead), since it asks for fewer seats than what the API server reserves for its PL at any time during the test. The test collects the server side APF metrics and client side latency metrics to calculate the "expected" concurrency utilization, and check it against the server-side "observed" concurrency utilization whether it falls within a margin of error. The margin of error is estimated by calculating the deviation of client request latency samples. We also check the server-side observed concurrency numbers to make sure they are not far off. 

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:
Attached the test log of test runs as examples.
[log2.txt](https://github.com/kubernetes/kubernetes/files/9112046/log2.txt)
[log1.txt](https://github.com/kubernetes/kubernetes/files/9112047/log1.txt)

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (MikeSpreitzer) [inline test/integration/apiserver/flowcontrol/concurrency_util_test.go:87]: Maybe comment where this code was copied from.
IIRC we thought it might be good to have a follow-on PR that refactors to remove the duplication.

**Review by MikeSpreitzer — COMMENTED** [review]
Looks pretty good.  I let one comment in-line.
It would be nice to have example `go test -v` output in either the git commit or the github PR.

**Comment by MikeSpreitzer** [comment]
@cyang49 : the pull-kubernetes-verify failure is saying that the exception list in hack/verify-prometheus-imports.sh needs to be augmented with your new file (you will see a sibling file already in there).

**Comment by MikeSpreitzer** [comment]
@cyang49 : the test itself is failing.  The proximal reason is that the attempts to read seat utilization metrics always return 0.  The test does not notice that failure or even log the values.  Those would be good things to fix in any case.

The root cause is #110104 , which replaced one of the metrics that the new test here is using.

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (cyang49) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:87]: Added a comment

**Comment by cyang49** [comment]
Attaching the log output of the updated PR
[log.txt](https://github.com/kubernetes/kubernetes/files/9123185/log.txt)

**Review by MikeSpreitzer — APPROVED** [review]
Looks good, thanks!

**Comment by apelisse** [comment]
Can I ask you tu run this test a few dozen times and see if it flakes?

**Comment by cyang49** [comment]
> Can I ask you tu run this test a few dozen times and see if it flakes?

@apelisse do you mean locally or in CI by triggering retests?

**Comment by apelisse** [comment]
Locally, I trust you to report the results properly :-)

**Comment by cyang49** [comment]
@apelisse I ran 10 times in a for loop and they all passed. Attaching the log files
[logs.zip](https://github.com/kubernetes/kubernetes/files/9213188/logs.zip)

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (lavalamp) [inline test/integration/apiserver/flowcontrol/concurrency_util_test.go:84]: Can we get a comment explaining what this test does and why it is needed?

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (lavalamp) [inline test/integration/apiserver/flowcontrol/concurrency_util_test.go:188]: These names are super long, that combined with repetition makes this code very hard to read.

I would declare a struct type and put these things inside it. I'd further make a sum/count pair struct.

Rename:
* earlier -> t0
* later -> t1
* RequestExecutionSecondsSum -> reqExecution.Sum
* RequestExecutionSecondsCount -> reqExecution.Count
* PLSeatUtilSum -> seat.Sum
* PLSeatUtilCount -> seat.Count

So the below math can say things like t0.reqExecution.Sum.

Put comments by the variable names that expand the terse names.

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (lavalamp) [inline test/integration/apiserver/flowcontrol/concurrency_util_test.go:110]: name change: plNoxu1?

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (lavalamp) [inline test/integration/apiserver/flowcontrol/concurrency_util_test.go:203]: I would make an additional type and function, so that all of these lines can look like `noxu1Avg := intervalAverage(t0, t1, plNoxu1.Name)` and the resulting type has members like `noxu1Avg.requestExecution` and `noxu1Avg.seats`.

(intervalAverage because what you are doing here is not just averaging)

The style as written makes it very easy to make a copy-paste error and very hard to visually detect one.

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (lavalamp) [inline test/integration/apiserver/flowcontrol/concurrency_util_test.go:211]: Make a stats struct to hold this pair of data; use SD or stdDev for the name of that value (not Sdev).

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (lavalamp) [inline test/integration/apiserver/flowcontrol/concurrency_util_test.go:242]: I'm very suspicious about fixing this via `Min`, if we graph this function it has a sharp inflection point which makes me suspicious of all other values too, not just the ones that went over the limit. I think we should not do this min and make sure the math tolerates large expectations, or find a better way to compute it.

Comment (lavalamp) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:242]: (I'd also like to see these variable names shortened, and/or temporaries computed, and/or a helper function called rather than repeat everything)

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (lavalamp) [inline test/integration/apiserver/flowcontrol/concurrency_util_test.go:246]: Please give a reference to the statistical test that we're doing here, or otherwise explain it?

Comment (lavalamp) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:246]: I am concerned that this is going to be too flaky, how many times have you run this test? Flakes definitely need to be < 1/100, ideally < 1/1000.

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (aojea) [inline test/integration/apiserver/flowcontrol/concurrency_util_test.go:151]: just curious, what is the `idx` param used for?

Comment (cyang49) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:151]: This is a remnant of a previous version. I will remove `streamReuqestsWithIndex` and use `streamRequests` (defined in concurrency_test.go instead.

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (cyang49) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:464]: You're right. This variable is used previously for debugging and not used anymore. I will remove it, and after the removal this struct is read only after initialization.

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (MikeSpreitzer) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:246]: This will be relevant: https://en.wikipedia.org/wiki/Error_function

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (MikeSpreitzer) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:242]: The test on line 137, making sure that availableSeats is larger than 4, ensures that this Min will have no effect for Noxu2.  The Min is here for Noxu2 only for the sake of uniformity; I would be OK with removing this Min.

For Noxu1, we have seen this Min make only a small difference.  The measurements that feed into this are not perfectly aligned (that is impossible Prometheus metrics) and have some other noise in them; the purpose of this Min is to cut out some divergence that comes from the noise.  Really, the comparisons that use these values are only checking that physics works as expected and these requests are not spending time in some totally unexpected way (recall that these checks are derived from equating throughput observed at two places, and that equality is not really optional).  @lavalamp, what graph did you have in mind?  I see three inputs that are not constant (`noxu1RequestExecutionSecondsAvg`, `noxu1ClientRequestLatencySecondsAvg`, and `availableSeats[priorityLevelNoxu1.Name]`).

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (lavalamp) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:246]: I would have expected a call to https://pkg.go.dev/math#Erf then, somewhere?

Comment (cyang49) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:246]: @lavalamp could you suggest some way to show the failure rate is low (i.e. not flaky) in CI? I can test in my environment thousands of times more, but I don't think it's meaningful if my environment is not as noisy as in CI

Comment (aojea) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:246]: https://github.com/kubernetes/kubernetes/pull/108553#issuecomment-1064871247

compile the test and run it with `stress -test.run Test..`

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (lavalamp) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:242]: Let me put it another way--

I expect comparisons of data that is noisy in these ways to have some bounded amount of fuzziness, but using min or max is effectively changing the observed data an arbitrary amount. I'd much rather compute the number as it happens, even if it is logically impossible for it to be that high, and tune the comparison(s) to have enough fudge factor to compensate.

Because if it's like 2% more than the max, that's fine, but if it is 200% more, that's a test bug and I want to know about it. Using min obscures the difference.

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (cyang49) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go]: This is resolved in the new commit

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (cyang49) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go]: this variable is removed in the new commit

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (cyang49) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go]: Code is refactored in the new version

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (cyang49) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:293]: I am collecting some data by removing the min and will report them here soon

Comment (cyang49) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:293]: @MikeSpreitzer @lavalamp 
I removed the Min and collected data from 50 tests and plotted the error, defined by `math.Abs(expected-observed)/expected`, and their ratio to the margin, defined by  `margin := 2 * math.Min(noxu1LatStats.cv, noxu2LatStats.cv)` (cv is [coefficient of variance](https://en.wikipedia.org/wiki/Coefficient_of_variation)). All 50 tests passed. The plots are as follows
client-latency-error
client-latency-error-to-margin-ratio

From the results it seems that there is some room for larger noise, since the error-to-margin ration is less than 0.25 in my experiments. 

Attaching also the logs
[Archive.zip](https://github.com/kubernetes/kubernetes/files/9245073/Archive.zip)

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (cyang49) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:297]: @aojea thanks for the tip. This revealed some failures from the checks of hard-coded error bound numbers. Will try to adjust them

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (lavalamp) [inline test/integration/apiserver/flowcontrol/concurrency_util_test.go:111]: use defer for unlocks.

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (lavalamp) [inline test/integration/apiserver/flowcontrol/concurrency_util_test.go:122]: shouldn't this lock?

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (lavalamp) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:297]: Thanks. You can perhaps add a bit more delay to the webhook to increase the signal, if you need to?

Also it's looking much more readable after the last push, thanks.

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (MikeSpreitzer) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:122]: I think that, at least, `type clientLatencyMeasurement` should have some comment(s) about concurrency.  And `func getStats` should obey what those comments say or elaborate on its exception.

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (MikeSpreitzer) [inline test/integration/apiserver/flowcontrol/concurrency_util_test.go:124]: Floating-point numbers generally have some noise in them, so it is possible that `ss < 0` --- which `Sqrt` will not like.

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (MikeSpreitzer) [inline test/integration/apiserver/flowcontrol/concurrency_util_test.go:132]: I would say "population standard deviation" (divide by N) to alert any speed reader who usually deals with "sample standard deviation" (divide by N-1); see, e.g., https://www.khanacademy.org/math/statistics-probability/summarizing-quantitative-data/variance-standard-deviation-sample/a/population-and-sample-standard-deviation-review for an exposition of the difference.

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (MikeSpreitzer) [inline test/integration/apiserver/flowcontrol/concurrency_util_test.go:151]: I think a better high level description would focus on the isolation.  Perhaps something like the following.

Remember also the convention that the comment on something begins with its name.

```
// TestConcurrencyIsolation tests the concurrency isolation between priority levels.
// The test defines two priority levels for this purpose, and corresponding flow schemas.
// To one priority level, this test sends many more concurrent requests than the configuration
// allows to execute at once, while sending fewer than allowed to the other priority level.
// The primary check is that the low flow gets all the seats it wants, but modulated by
// recognizing that there are uncontrolled overheads in the system.
```

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (MikeSpreitzer) [inline test/integration/apiserver/flowcontrol/concurrency_util_test.go:264]: typo

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (MikeSpreitzer) [inline test/integration/apiserver/flowcontrol/concurrency_util_test.go:233]: This outer layer of map structure is unnecessary complexity.
It would be more direct to just use two variables of type `metricSnapshot`, perhaps named `snapshot0` and `snapshot1`.

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (MikeSpreitzer) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:122]: The lack of locking here is OK because this func is called after the concurrent goroutines stop updating the same data.  But there should be commentary (as I suggested above) so that readers like @lavalamp are guided to see the validity of the code.

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (MikeSpreitzer) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:297]: To answer the original question, the idea is very basic.  We know there are uncontrolled overheads that introduce noise into the system.  The coefficient of variation --- that is, standard deviation divided by mean --- for a class of traffic is a characterization of all the noise that applied to that class.  We found that class 1 generally had a much bigger CV than class 2.  This makes sense, because class 1 probes more behavior --- the waiting in queues.  So we take the minimum of the two as an indicator of the relative amount of noise that comes from all the other behavior.

We picked two standard deviations because it is commonly used as a discriminator between common and uncommon.  We do not need to call the error function at runtime, because we are using a fixed multiplier of the standard deviation.  Using 2 as that multiplier gets the result that we expect about 95.4% of trials to fall within the bound.  A multiplier of 3 would expect 99.7% to fall within the bound.

Of course, the above is assuming that the noise is normally distributed, which we are not really justified in assuming.  So all this is really random precision.  We just wanted some way to make the test adjust to the noise it experienced.

**Comment by MikeSpreitzer** [comment]
@lavalamp: while it is necessary that @cyang49 get this test to run reliably on his development machine, that would not convince me that it will be reliable in our CI environment (which is much more variable in its levels of concurrent stresses).  Is there anything we could do to get more confidence of the latter?

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (cyang49) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:124]: @MikeSpreitzer what's the better way to deal with that problem? Should I clip a negative value and make it 0.0?

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (cyang49) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:297]: I'll incorporate Mike's comment into the code comment as the explanation on the margin of error calculation

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (MikeSpreitzer) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:126]: yes

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (MikeSpreitzer) [inline test/integration/apiserver/flowcontrol/concurrency_util_test.go:144]: s/snapshot0 metricSnapshot, snapshot1 metricSnapshot/snapshot0, snapshot1 metricSnapshot/

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (MikeSpreitzer) [inline test/integration/apiserver/flowcontrol/concurrency_util_test.go]: I was not trying to suggest deleting this text.

But I would revise a bit.  Perhaps something like the following.

```
// This test differs from TestPriorityLevelIsolation since TestPriorityLevelIsolation checks throughput instead
// of concurrency. In order to mitigate the effects of system noise, authorization webhook is used to artificially
// increase request execution time to make the system noise relatively insignificant.
//
// Secondarily, this test also checks the observed seat utilizations against values derived from expecting that
// the throughput observed by the client equals the execution throughput observed by the server.
//
// This test recognizes that there is noise in the measurements coming from uncontrolled overheads
// and unsynchronized reading of related quantities.  This test takes as a relative error margin 2 times
// the smaller (of the two traffic classes) coefficient of variation of the client-observed latency.
```

**Comment by lavalamp** [comment]
> @lavalamp: while it is necessary that @cyang49 get this test to run reliably on his development machine, that would not convince me that it will be reliable in our CI environment (which is much more variable in its levels of concurrent stresses). Is there anything we could do to get more confidence of the latter?

Running with stress on one's dev machine (as @cyang49 has done) seems to be pretty good. We're only looking for 2 or 3 nines here, not 5 :)

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (cyang49) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go]: I stress tested the latest version
```
$ stress ./flowcontrol.test -test.run=TestConcurrencyIsolation
5s: 0 runs so far, 0 failures
10s: 0 runs so far, 0 failures
15s: 0 runs so far, 0 failures
20s: 0 runs so far, 0 failures
25s: 0 runs so far, 0 failures
30s: 0 runs so far, 0 failures
35s: 16 runs so far, 0 failures
40s: 16 runs so far, 0 failures
45s: 16 runs so far, 0 failures
50s: 16 runs so far, 0 failures
55s: 16 runs so far, 0 failures
1m0s: 18 runs so far, 0 failures
...
33m55s: 1059 runs so far, 0 failures
34m0s: 1061 runs so far, 0 failures
34m5s: 1064 runs so far, 0 failures
34m10s: 1066 runs so far, 0 failures
34m15s: 1069 runs so far, 0 failures
```

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (cyang49) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go]: Changed to suggested in the new version

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (lavalamp) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:122]: I suspected as much, but I think there's no harm in locking here and it would prove that a stray write at least won't be only partially counted?

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (cyang49) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go]: Incorporated in the new version

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (cyang49) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:122]: Agree and added lock/unlock

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (lavalamp) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:126]: I'm not convinced it's possible to construct a series of measurements that have a net quantization error in that direction, but I guess I'm also not convinced it isn't possible, so I guess I agree with adding the check. IEEE floats are honestly pretty weird animals.

**Comment by lavalamp** [comment]
I'm gonna be OOO for a bit and I think this is very close now, so I'll approve and Mike can give the last LGTM when he thinks it's done?

/approve

I support adding this test to the release but at this point you have to convince someone on the release team, I think.

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (cyang49) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:126]: Resolved in the new version

Comment (MikeSpreitzer) [reply test/integration/apiserver/flowcontrol/concurrency_util_test.go:126]: BTW, it is not just a matter of quantization error.  With floating point numbers, addition is not even associative.

File: test/integration/apiserver/flowcontrol/concurrency_util_test.go
Comment (MikeSpreitzer) [inline test/integration/apiserver/flowcontrol/concurrency_util_test.go:319]: I myself would compute the relative error once.
This is a really minor style thing.

**Review by MikeSpreitzer — APPROVED** [review]
Noted a couple minor issues in-line; can be addressed in follow-up if you like.

/LGTM

**Review by MikeSpreitzer — APPROVED** [review]
/lgtm
Thanks!

**Pull request merged**
